opt_einsum/sharing.py

```diff
@@ -4,10 +4,37 @@
 import functools
 import numbers
 from collections import Counter
+import threading
 
 from .parser import alpha_canonicalize, parse_einsum_input
 
-_SHARING_STACK = []
+
+_SHARING_STACK = {}
+
+
+def currently_sharing():
+    """Check if we are currently sharing a cache -- thread specific.
+    """
+    return threading.get_ident() in _SHARING_STACK
```

Comment:
Hmm in Python 2, `threading.get_ident` does not exist

Reply:
Yes, just realized that! I'll have a quick look for an alternative.